### PR TITLE
secondary-storage: delete backedup snapshot dir on delete

### LINKED ISSUE #7516: Deleting snapshots leaves empty directories
CloudStack doesn't delete snapshot directories when snapshots are deleted. This leaves behind empty directories on the secondary storage's `snapshots` directory.

As a side-effect of this, very large number of empty directories can cause template/ISO upload failure due to SSVM checking these directories for capacity/limit checks.

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Snapshots, Secondary Storage, SSVM
~~~

##### CLOUDSTACK VERSION

~~~
4.17.2.0, 4.18.0.0 (possibly by design all versions of ACS)
~~~

##### CONFIGURATION

Advance zone environment, tested and reproduce this against VMware 6.5u2, 6.7u3 and EL8/KVM.

##### STEPS TO REPRODUCE

~~~
1. Deploy a VM and create volume snapshot of its disks.
2. Delete the volume snapshot once they are backed up.
3. On the secondary storage, the snapshots are stored at the path convention: `snapshots/<volume ID>/<snapshot ID>/<snapshots files or folders>`. Before deletion of the snapshots, you should be able to see snapshots files in the snapshots directory for the specific volume/snapshot.
4. Post-deletion of the snapshots, we see empty directories which aren't removed.
~~~

##### EXPECTED RESULTS

~~~
Empty directories aren't kept, but removed
~~~

##### ACTUAL RESULTS

~~~
snapshot directories had several empty directories for removed snapshots
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #7516

When a backed-up snapshot is deleted and the snapshot file is present in the same name directory (probably only for VMware), the whole directory can be deleted.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

**Tested deleting a backedup volume snapshot on VMware**
Logs:
```
2023-05-12 09:45:49,879 DEBUG [cloud.agent.Agent] (agentRequest-Handler-4:null) Request:Seq 5-731553464470994956:  { Cmd , MgmtId: 32988117074588, via: 5, Ver: v1, Flags: 100011, [{"org.apache.cloudstack.storage.command.DeleteCommand":{"data":{"org.apache.cloudstack.storage.to.SnapshotObjectTO":{"path":"snapshots/2/10/900f3dd1-5be2-4036-ae08-b93548d9e5c8/900f3dd1-5be2-4036-ae08-b93548d9e5c8","volume":{"uuid":"56730834-50ec-45be-b1e0-2040a271a1c3","volumeType":"ROOT","dataStore":{"org.apache.cloudstack.storage.to.PrimaryDataStoreTO":{"uuid":"e13a6bdb-50ff-30d9-b446-653989234532","name":"ref-trl-4867-v-M7-abhishek-kumar-esxi-pri2","id":"2","poolType":"NetworkFilesystem","host":"10.0.32.4","path":"/acs/primary/ref-trl-4867-v-M7-abhishek-kumar/ref-trl-4867-v-M7-abhishek-kumar-esxi-pri2","port":"2049","url":"NetworkFilesystem://10.0.32.4/acs/primary/ref-trl-4867-v-M7-abhishek-kumar/ref-trl-4867-v-M7-abhishek-kumar-esxi-pri2/?ROLE=Primary&STOREUUID=e13a6bdb-50ff-30d9-b446-653989234532","isManaged":"false"}},"name":"ROOT-10","size":"(2.00 GB) 2147483648","path":"ROOT-10","volumeId":"10","vmName":"i-2-10-VM","accountId":"2","chainInfo":"{"diskDeviceBusName":"ide0:1","diskChain":["[e13a6bdb50ff30d9b446653989234532] i-2-10-VM/ROOT-10.vmdk","[e13a6bdb50ff30d9b446653989234532] 85ae42cf969a3bd99d767590746a5209/85ae42cf969a3bd99d767590746a5209.vmdk"]}","format":"OVA","provisioningType":"THIN","poolId":"2","id":"10","deviceId":"0","hypervisorType":"VMware","directDownload":"false","deployAsIs":"false"},"dataStore":{"com.cloud.agent.api.to.NfsTO":{"_url":"NFS://10.0.32.4/acs/secondary/ref-trl-4867-v-M7-abhishek-kumar/ref-trl-4867-v-M7-abhishek-kumar-sec1","_role":"Image"}},"vmName":"i-2-10-VM","name":"VM-5128d690-2204-415a-85e8-fb6516d1974f_ROOT-10_20230512094317","hypervisorType":"VMware","id":"6","quiescevm":"false","physicalSize":"0"}},"wait":"0","bypassHostMaintenance":"false"}}] }
2023-05-12 09:45:49,883 DEBUG [cloud.agent.Agent] (agentRequest-Handler-4:null) Processing command: org.apache.cloudstack.storage.command.DeleteCommand
2023-05-12 09:45:49,886 INFO  [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Determined host 10.0.32.4 corresponds to IP 10.0.32.4
2023-05-12 09:45:49,886 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Mounting device with nfs-style path of 10.0.32.4:/acs/secondary/ref-trl-4867-v-M7-abhishek-kumar/ref-trl-4867-v-M7-abhishek-kumar-sec1
2023-05-12 09:45:49,887 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) mount NFS://10.0.32.4/acs/secondary/ref-trl-4867-v-M7-abhishek-kumar/ref-trl-4867-v-M7-abhishek-kumar-sec1 on /mnt/SecStorage/18061267-f04b-31f5-9b46-f508765a8cef nfsVersion=null
2023-05-12 09:45:49,887 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) making available /mnt/SecStorage/18061267-f04b-31f5-9b46-f508765a8cef on NFS://10.0.32.4/acs/secondary/ref-trl-4867-v-M7-abhishek-kumar/ref-trl-4867-v-M7-abhishek-kumar-sec1
2023-05-12 09:45:49,887 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) local folder for mount will be /mnt/SecStorage/18061267-f04b-31f5-9b46-f508765a8cef
2023-05-12 09:45:49,888 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Executing: mount 
2023-05-12 09:45:49,893 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Executing while with timeout : 1440000
2023-05-12 09:45:49,894 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Execution is successful.
2023-05-12 09:45:49,894 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Some device already mounted at /mnt/SecStorage/18061267-f04b-31f5-9b46-f508765a8cef, no need to mount NFS://10.0.32.4/acs/secondary/ref-trl-4867-v-M7-abhishek-kumar/ref-trl-4867-v-M7-abhishek-kumar-sec1
2023-05-12 09:45:49,909 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Snapshot file 900f3dd1-5be2-4036-ae08-b93548d9e5c8 is present in the same name directory /mnt/SecStorage/18061267-f04b-31f5-9b46-f508765a8cef/snapshots/2/10/900f3dd1-5be2-4036-ae08-b93548d9e5c8. Deleting the directory
2023-05-12 09:45:49,909 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Executing: /bin/bash -c rm -rf /mnt/SecStorage/18061267-f04b-31f5-9b46-f508765a8cef/snapshots/2/10/900f3dd1-5be2-4036-ae08-b93548d9e5c8 
2023-05-12 09:45:49,911 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Executing while with timeout : 3600000
2023-05-12 09:45:50,014 DEBUG [storage.resource.NfsSecondaryStorageResource] (agentRequest-Handler-4:null) Execution is successful.
```

Storage
before deletion:
```
[root@nfs snapshots]# tree
.
└── 2
    ├── 10
    │   ├── 1039daff-ef99-41df-bffd-ee345490d42b
    │   ├── 52794c72-b196-4de2-aaa8-3a0132806a15
    │   ├── 900f3dd1-5be2-4036-ae08-b93548d9e5c8
    │   │   ├── 900f3dd1-5be2-4036-ae08-b93548d9e5c8-disk0.vmdk
    │   │   ├── 900f3dd1-5be2-4036-ae08-b93548d9e5c8-disk1.nvram
    │   │   └── 900f3dd1-5be2-4036-ae08-b93548d9e5c8.ovf
    │   ├── bd232cab-baa1-4ac5-9b10-bbc48c2bc312
    │   └── fd57cb44-0559-490c-998d-5313eec52cef
    │       ├── fd57cb44-0559-490c-998d-5313eec52cef-disk0.vmdk
    │       ├── fd57cb44-0559-490c-998d-5313eec52cef-disk1.nvram
    │       └── fd57cb44-0559-490c-998d-5313eec52cef.ovf
    └── 9
        └── 91fa77c2-15cc-412d-9358-b1178730ce43

9 directories, 6 files
```

after deletion:
```
[root@sl-nestednfs snapshots]# tree
.
└── 2
    ├── 10
    │   ├── 1039daff-ef99-41df-bffd-ee345490d42b
    │   ├── 52794c72-b196-4de2-aaa8-3a0132806a15
    │   ├── bd232cab-baa1-4ac5-9b10-bbc48c2bc312
    │   └── fd57cb44-0559-490c-998d-5313eec52cef
    │       ├── fd57cb44-0559-490c-998d-5313eec52cef-disk0.vmdk
    │       ├── fd57cb44-0559-490c-998d-5313eec52cef-disk1.nvram
    │       └── fd57cb44-0559-490c-998d-5313eec52cef.ovf
    └── 9
        └── 91fa77c2-15cc-412d-9358-b1178730ce43

8 directories, 3 files
```

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 1]
```diff
@@ -1997,6 +1997,17 @@ private void savePostUploadPSK(String psk) {
...
+    protected String getSnapshotFilepathForDelete(String path, String snapshotName) {
+        String finalPath = path + "/*" + snapshotName + "*";
+        if (!path.endsWith(snapshotName)) {
+            return finalPath;
+        }
+        if (s_logger.isDebugEnabled()) {
+            s_logger.debug(String.format("Snapshot file %s is present in the same name directory %s. Deleting the directory", snapshotName, path));
+        }
+        return path;
+    }
```
Comment (weizhouapache) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:2009]: @shwstppr 
would it be safer to check if the path is a directory and there is no files in the directory before deletion ?

Comment (shwstppr) [reply services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:2009]: @weizhouapache directory check be added but we always expect it to be a directory anyway as we append the `/*SNAPSHOT_NAME*

And there will always be files (I checked  VMware, Xen and KVM. Snapshots are placed inside the same name folder **only for VMWare**). Example

```
[root@sl-nestednfs snapshots]# tree 2/10/
2/10/
└── fd57cb44-0559-490c-998d-5313eec52cef
    ├── fd57cb44-0559-490c-998d-5313eec52cef-disk0.vmdk
    ├── fd57cb44-0559-490c-998d-5313eec52cef-disk1.nvram
    └── fd57cb44-0559-490c-998d-5313eec52cef.ovf
```

Comment (rohityadavcloud) [reply services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:2009]: I suppose we need to cover both cases, to check if we're deleting files inside a folder (like in vmware), but also for kvm where snapshots arne't in a folder. XenServer/XCP-ng we need to check.

File: services/secondary-storage/server/src/test/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResourceTest.java [hunk 2]
```diff
@@ -91,4 +91,21 @@ public void testCleanupStagingNfs() throws Exception{
...
 
     }
+
+    private void performGetSnapshotFilepathForDeleteTest(String expected, String path, String name) {
+        Assert.assertEquals("Incorrect resultant snapshot delete path", expected, resource.getSnapshotFilepathForDelete(path, name));
+    }
+
+    @Test
+    public void testGetSnapshotFilepathForDelete() {
+        performGetSnapshotFilepathForDeleteTest("/snapshots/2/10/somename",
```
Comment (rohityadavcloud) [inline services/secondary-storage/server/src/test/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResourceTest.java:101]: 👍

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan LLtest rocky8 vmware-67u3

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 3]
```diff
@@ -1997,6 +1997,17 @@ private void savePostUploadPSK(String psk) {
         }
     }
 
+    protected String getSnapshotFilepathForDelete(String path, String snapshotName) {
+        String finalPath = path + "/*" + snapshotName + "*";
+        if (!path.endsWith(snapshotName)) {
+            return finalPath;
+        }
```
Comment (DaanHoogland) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:2004]:
```suggestion
        if (!path.endsWith(snapshotName)) {
            return path + "/*" + snapshotName + "*";
        }
```

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java (hunk 1 above)
Comment (DaanHoogland) [reply services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:2009]: @weizhouapache @shwstppr , we are dealing with a `String`, not a `File`/directory. I think this is ok, in this place. checking for existence should not happen in this method.

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by shwstppr** [comment]
@blueorangutan LLpackage

**Comment by shwstppr** [comment]
@blueorangutan LLtest matrix

**Comment by rohityadavcloud** [comment]
@blueorangutan test matrix

**Pull request merged**